### PR TITLE
Empty collection request optimization

### PR DESCRIPTION
**Pull request by AliveMen**
used isNullOrEmpty
minor reformat

### CODE REVIEW

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 1]
```diff
@@ -183,28 +183,33 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
 
         public virtual ShoppingCartEntity[] GetShoppingCartsByIds(string[] ids, string responseGroup = null)
         {
-            var carts = ShoppingCarts.Where(x => ids.Contains(x.Id)).ToArray();
+            var result = new ShoppingCartEntity[0];
```
Comment (artem-dudarev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: It's not enough. If there are no carts you will still perform redundant requests.
Also there is no need to create an empty array if ids is not empty.
```csharp
ShoppingCartEntity[] result;

if (ids.IsNullOrEmpty())
{
    result = new ShoppingCartEntity[0];
}
else
{
    result = ShoppingCarts.Where(x => ids.Contains(x.Id)).ToArray();

    if (result.Any())
    {
        ids = result.Select(x => x.Id).ToArray();
        ...
    }
}

return result;
```

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 2]
```diff
@@ -183,28 +183,41 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+            {
+                result = ShoppingCarts.Where(x => ids.Contains(x.Id)).ToArray();
 
-            var shipmentIds = Shipments.Include(x => x.Items).Where(x => ids.Contains(x.ShoppingCartId)).ToArray().Select(x => x.Id).ToArray();
-            var shipmentTaxDetails = TaxDetails.Where(x => shipmentIds.Contains(x.ShipmentId)).ToArray();
-            var shipmentDiscounts = Discounts.Where(x => shipmentIds.Contains(x.ShipmentId)).ToArray();
-            var shipmentAddresses = Addresses.Where(x => shipmentIds.Contains(x.ShipmentId)).ToArray();
+                if (result.Any())
+                {
+                    ids = result.Select(x => x.Id).ToArray();
```
Comment (artem-dudarev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Separate with empty line from the following code

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 3]
```diff
@@ -183,28 +183,41 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+                    var cartTaxDetails = TaxDetails.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartDiscounts = Discounts.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartAddresses = Addresses.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartCoupons = Coupons.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+
+                    var paymentIds = Payments.Include(x => x.Addresses)
```
Comment (tatarincev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Need to add the same IsNullOrEmpty check for paymentIds

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 4]
```diff
@@ -183,28 +183,41 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+                        .Where(x => ids.Contains(x.ShoppingCartId)).ToArray().Select(x => x.Id).ToArray();
+                    var paymentTaxDetails = TaxDetails.Where(x => paymentIds.Contains(x.PaymentId)).ToArray();
+                    var paymentDiscounts = Discounts.Where(x => paymentIds.Contains(x.PaymentId)).ToArray();
+
+                    var lineItemIds = LineItems.Where(x => ids.Contains(x.ShoppingCartId)).ToArray().Select(x => x.Id).ToArray();
```
Comment (tatarincev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Need to add the same IsNullOrEmpty check for lineItemIds

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 5]
```diff
@@ -183,28 +183,41 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+                    var lineItemsTaxDetails = TaxDetails.Where(x => lineItemIds.Contains(x.LineItemId)).ToArray();
+                    var lineItemsDiscounts = Discounts.Where(x => lineItemIds.Contains(x.LineItemId)).ToArray();
+
+                    var shipmentIds = Shipments.Include(x => x.Items).Where(x => ids.Contains(x.ShoppingCartId)).ToArray().Select(x => x.Id).ToArray();
```
Comment (tatarincev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Need to add the same IsNullOrEmpty check for shipmentIds

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs (hunk 2 above)
Comment (yecli) [reply VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Fixed

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs (hunk 3 above)
Comment (yecli) [reply VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Fixed

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs (hunk 4 above)
Comment (yecli) [reply VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Fixed

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs (hunk 5 above)
Comment (yecli) [reply VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Fixed

File: VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs [hunk 6]
```diff
@@ -183,28 +184,52 @@ protected override void OnModelCreating(DbModelBuilder modelBuilder)
...
+                {
+                    ids = result.Select(x => x.Id).ToArray();
+
+                    var cartTaxDetails = TaxDetails.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartDiscounts = Discounts.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartAddresses = Addresses.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+                    var cartCoupons = Coupons.Where(x => ids.Contains(x.ShoppingCartId)).ToArray();
+
+                    var paymentIds = Payments.Include(x => x.Addresses)
+                        .Where(x => ids.Contains(x.ShoppingCartId)).ToArray().Select(x => x.Id).ToArray();
```
Comment (artem-dudarev) [inline VirtoCommerce.CartModule.Data/Repositories/CartRepositoryImpl.cs]: Move `.ToArray` and `.Select` to separate lines

**Comment by artem-dudarev** [comment]
@tatarincev @yecli Please merge and release this PR.

**Pull request merged**